record/record_test.go

```diff
@@ -312,3 +313,145 @@ func TestStaleWriter(t *testing.T) {
+	}
+
+	r1Data, err := ioutil.ReadAll(r1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !bytes.Equal(r1Data, records[0]) {
+		t.Fatal("Unexpected output in record 1's data.")
+	}
+
+	// This record should present problems since the checksum is wrong.
+	_, err = r.Next()
+	if err == nil {
+		t.Fatal("Expected an error while reading a corrupted record.")
+	}
+
+	if !strings.Contains(err.Error(), "checksum mismatch") {
+		t.Fatalf("Unexpected error returned: %s", err)
+	}
+
+	// Now attempt to recover from the checksum error we intentionally caused.
+	r.Recover()
+
+	currentOffset, err := underlyingReader.Seek(0, os.SEEK_CUR)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if currentOffset != blockSize*2 {
+		t.Fatalf("After calling r.Recover we expected the underlying cursor offset to be %d, but received %d", blockSize*2, currentOffset)
+	}
+
+	r3, err := r.Next()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	data, err := ioutil.ReadAll(r3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !bytes.Equal(data, records[2]) {
+		t.Fatal("Unexpected output in recovered data.")
+	}
+}
+
+func TestComplexReadRecovery(t *testing.T) {
+	// The first record will be blockSize * 3 bytes long. Since each block has a
+	// 6 byte header, the first record will roll over into 4 blocks.
+	records := [][]byte{
+		[]byte(strings.Repeat("a", blockSize*3)),
+		[]byte(strings.Repeat("b", blockSize-headerSize)),
+		[]byte(strings.Repeat("c", blockSize-headerSize)),
+	}
+	buf := new(bytes.Buffer)
+
+	w := NewWriter(buf)
+	for i := 0; i < 3; i++ {
+		wRec, err := w.Next()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if _, err = wRec.Write(records[i]); err != nil {
+			t.Fatal(err)
+		}
+	}
+	w.Close() // Finalize the last record.
+	rawBufSlice := buf.Bytes()
+
+	// Now corrupt the checksum for the portion of the first record that exists in the 4th block.
+	rawBufSlice[blockSize*3+0] = 0xef
+	rawBufSlice[blockSize*3+1] = 0xbe
+	rawBufSlice[blockSize*3+2] = 0xad
+	rawBufSlice[blockSize*3+3] = 0xde
+	r := NewReader(bytes.NewReader(rawBufSlice))
+
+	// The first record should fail, but only when we read deeper beyond the first block.
+	if r1, err := r.Next(); err == nil {
+		// err below should reference a checksum mismatch.
+		_, err := ioutil.ReadAll(r1)
+		if err == nil {
+			t.Fatal(err)
+		}
+		if !strings.Contains(err.Error(), "checksum mismatch") {
+			t.Fatalf("Unexpected error returned: %s", err)
+		}
+	}
+
+	// Recover from the checksum mismatch
+	r.Recover()
+
+	// All of the data in the second record is lost because the first record shared a partial
+	// block with it. The second record also overlapped into the block with the third record.
+	// Recovery was able to jump to that block, skipping over the end of the second record and
+	// start parsing the third record which we verify below.
+	if r3, err := r.Next(); err != nil {
+		t.Fatal(err)
+	}
+
+	r3Data, _ := ioutil.ReadAll(r3)
+	if !bytes.Equal(r3Data, records[2]) {
+		t.Fatal("Unexpected output in record 3's data.")
+	}
+}
```

Comment:
Can I get a test where the corruption happens in the last block, not a middle block?